### PR TITLE
add support for calculating CVSS score from the CVSS vector

### PR DESCRIPTION
**Pull request by ziadhany**
Reference: #713

Signed-off-by: Ziad <ziadhany2016@gmail.com>

### CODE REVIEW

**Review by pombredanne — CHANGES_REQUESTED** [review]
Looking good... Can you add some tests?

**Review by pombredanne — CHANGES_REQUESTED** [review]
LGTM! :+1: 
@ziadhany do you mind to rebase on the latest code and also regen the migration on the latest code?
Also there is a need for a data migrations to migrate the current CVSS vector to this new approach

**Comment by pombredanne** [comment]
re: 
>  Also there is a need for a data migrations to migrate the current CVSS vector to this new approach

CVSSV31,
        CVSSV31_VECTOR,

say we have two rows with these data:
- scoring_system: CVSSV31_VECTOR value:"CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:L/A:H"
- scoring_system: CVSSV31 value: "8.6"

After the migration(s) I would like to see only one record:
- scoring_system: CVSSV31 value: "8.6" scoring_elements: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:L/A:H"

This is best done in three steps:
1. model schema update adding scoring_elements
2. data migration that searches and adds or computes the score for CVSS severities and add these to one of the two records
2.1 one way: 
starting from this record - scoring_system: CVSSV31_VECTOR value:"CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:L/A:H"
update the record this - update scoring_system: CVSSV31, copy value to value to scoring_elements, compute value
2.2 other way: group pairs of records and merge the two in one of the two records
3. data migration to remove the extra record

File: vulnerabilities/models.py [hunk 1]
```diff
@@ -369,7 +367,27 @@ class VulnerabilitySeverity(models.Model):
...
+
+    scoring_elements = models.CharField(
+        max_length=100,
+        null=True,
+        help_text="Supporting a scoring elements as a string "
+        "For example a CVSS vector Important, High, Medium ,Low."
+        "Typically used to compute the value",
+    )
+
+    def save(self, *args, **kwargs):
```
Comment (pombredanne) [inline vulnerabilities/models.py:383]: IMHO we should not save here

**Review by pombredanne — CHANGES_REQUESTED** [review]

File: vulnerabilities/models.py (hunk 1 above)
Comment (TG1999) [reply vulnerabilities/models.py:383]: I think we should compute it in `severity_system.py`

File: vulnerabilities/migrations/0030_vulnerabilityseverity _rm_extra_records_swap.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+from django.db import migrations
+
+
+def remove_extra_rows(apps, schema_editor):
+    Vuln_severity = apps.get_model('vulnerabilities', 'VulnerabilitySeverity')
+    for vuln_severity in Vuln_severity.objects.all():
```
Comment (TG1999) [inline vulnerabilities/migrations/0030_vulnerabilityseverity _rm_extra_records_swap.py:6]: @ziadhany please check 0025 to see how we are removing duplicates

File: vulnerabilities/migrations/0030_vulnerabilityserverity_compute_score.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from django.db import migrations
+
+from vulnerabilities.severity_systems import SCORING_SYSTEMS
+
+
+def compute_score(apps, schema_editor):
+    Vuln_severity = apps.get_model('vulnerabilities', 'VulnerabilitySeverity')
+    for vuln_severity in Vuln_severity.objects.all():
```
Comment (pombredanne) [inline vulnerabilities/migrations/0030_vulnerabilityserverity_compute_score.py:8]: Do not iterate all, instead you only care for the cvss so filter on them.

File: vulnerabilities/migrations/0030_vulnerabilityserverity_compute_score.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        if vuln_severity.value and not vuln_severity.scoring_elements:
+            try:
+                vuln_severity.scoring_elements = SCORING_SYSTEMS[vuln_severity.scoring_system].as_score(vuln_severity.value)
```
Comment (pombredanne) [inline vulnerabilities/migrations/0030_vulnerabilityserverity_compute_score.py:11]: What if we compute a score that's not the same the one we had before?
We should find a way to do what I suggest as a possibility e.g, " 2.2 other way: group pairs of records and merge the two in one of the two records" and ensure that we do not end up computing a score that's not the score we have stored

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks for the update! please see some comments inline.

**Comment by pombredanne** [comment]
@ziadhany I pushed these updates:
1. merge the latest main branch including handling overlapping migration numbers
2. update the the migration(s) to use a bulk update/bulk delete approach
3. propagate the use of scoring_elements everywhere including importers, tests, API, etc.
4. added some cosmetic improvements to regen tests fixtures easily

The new migration(s) handle all possible duplicates cases this way:

- keep a mapping of severities keyed by the "unique together" values of the model
- keep a set of severity ids to delete
- for each CVSS-related severity:
  - if this is a vector:
    - update this to move value to scoring elements and compute score in value and use plain cvss scoring system
  - if it does not exist in the mapping
    - add it to this mapping
  - else
    - if needed merge/add vector to the existing severity
    - mark this severity for deletion
- bulk delete the set of ids 
- bulk update severities mapping
 

I need to test run it on the whole DB to validate this is correct

**Comment by pombredanne** [comment]
@ziadhany This is essentially the same approach as yours, but folded in a single migration with bulk updates

**Review by pombredanne — APPROVED** [review]

**Comment by ziadhany** [comment]
LGTM ,  I tested it with my local database

**Comment by TG1999** [comment]
@ziadhany please add CHANGEOLG for this PR

**Pull request merged**